Add lesson: Debugging with ChatGPT (#2)
Ada-Developers-Academy/independent-ai-topics: 03-debugging-w-chatgpt/debugging-w-chatgpt.md

# Debugging with ChatGPT

## Goals

> We have a number of tools in our belt for debugging and understanding code, but there’s always room for more! This lesson is about what we can do with LLMs and generative AI tools like ChatGPT if we’ve used all the debugging tools at our disposal and are still feeling stuck or unsure of what a piece of code is doing.

Reviewer: What do you think about adding 'other'? --> "if we’ve used all other debugging tools at our disposal"

> We have a number of tools in our belt for debugging and understanding code, but there’s always room for more! This lesson is about what we can do with LLMs and generative AI tools like ChatGPT if we’ve used all the debugging tools at our disposal and are still feeling stuck or unsure of what a piece of code is doing.

Author: ```suggestion
We have a number of tools in our belt for debugging and understanding code, but there’s always room for more! This lesson is about what we can do with LLMs and generative AI tools like ChatGPT if we’ve used all other debugging tools at our disposal and are still feeling stuck or unsure of what a piece of code is doing.
```

Our goals for this lesson are to:

> - Learn how to effectively share code and context around code with ChatGPT

Reviewer: Is this meant to be "Learn how to effectively share code and context with ChatGPT" ?

> - Learn how to effectively share code and context around code with ChatGPT

Author: ```suggestion
- Learn how to effectively share code and context with ChatGPT
```

- Use ChatGPT in conjunction with our other debugging skills 

> - Examine how to evaluate the explanations ChatGPT gives for correctness

Reviewer: What about "Examine how to evaluate ChatGPT provided explanations for correctness" ? Maybe I'm being nitpicky.

Author: I like that, it feels more concise.

> - Examine how to evaluate the explanations ChatGPT gives for correctness

Author: ```suggestion
- Examine how to evaluate ChatGPT provided explanations for correctness
```

> ## Caveats & Concerns

Reviewer: What about using a warning callout here to draw attention to this section?

Author: I was wondering if it was a bit long to place in a warning callout, I'll make that swap because I really want to make sure folks take this in.

First and foremost, we want to note that all of our debugging skills that we learned before are still necessary and helpful, even with generative AI tools. In many instances, we’ll be able to find and fix a bug faster by reading through the code or adding a couple break points than it would take to craft and iterate on a prompt to get an explanation. Furthermore, because tools like ChatGPT can fabricate information and are not running code themselves to ensure what they suggest works, we need to keep up our debugging and reasoning skills to evaluate what ChatGPT shares with us to see if it 1) makes sense and 2) truly addresses the initial problem.

Second, and very importantly: depending on what it is you’re working on, you may not be allowed to send the code to ChatGPT or other generative AI tools. Enterprise solutions are being created and improved that allow companies to feed their proprietary code into an LLM-based tool without the code being used for further training or made in some way available to other people, but this is not the case for most publicly available LLMs and generative AI tools at this time. 

> Be very careful about what you share with LLMs - personal projects that you have complete control over and don’t mind the code being shared with the companies training these tools are great candidates for using ChatGPT to explain concepts and debug with. Any project you don’t have full ownership over should be discussed and approved by other stakeholders before code is shared with generative AI tools of any kind.

Reviewer: This sentence tripped me up a little bit. What about something like "An example of code that you could submit to ChatGPT to receive an explanation of concepts or debug with would be a personal project that you have complete control over."

> Be very careful about what you share with LLMs - personal projects that you have complete control over and don’t mind the code being shared with the companies training these tools are great candidates for using ChatGPT to explain concepts and debug with. Any project you don’t have full ownership over should be discussed and approved by other stakeholders before code is shared with generative AI tools of any kind.

Author: ```suggestion
Be very careful about what you share with LLMs - an example of code that you could submit to ChatGPT to receive an explanation of concepts or debug with would be a personal project that you have complete control over. Any project which you don’t have full ownership over should be discussed and approved by other stakeholders before code is shared with generative AI tools of any kind.
```

## Tips for working with code in LLMs

### Setting the Stage

> It’s helpful to set the stage first when working with LLM tools like ChatGPT for code. We can do this by asking ChatGPT to act as an expert in a particular language, helping you to debug. If we know exactly what problem we need help with we can include more details in the prompt, but a starting prompt could look like:

Reviewer: ```suggestion
It’s helpful to set the stage first when working with LLM tools like ChatGPT while coding. We can do this by asking ChatGPT to act as an expert in a particular language which can help your debugging. 
```

> It’s helpful to set the stage first when working with LLM tools like ChatGPT for code. We can do this by asking ChatGPT to act as an expert in a particular language, helping you to debug. If we know exactly what problem we need help with we can include more details in the prompt, but a starting prompt could look like:

Author: ```suggestion
It’s helpful to set the stage first when working with LLM tools like ChatGPT while coding. We can do this by prompting ChatGPT to "act as an expert in a particular language, helping you to debug". If we know exactly what problem we need help with we can include more details in the prompt, but a starting prompt could look like:
```

> You are a python expert who will help me debug python code.

### Gathering Information

> Adding in ChatGPT doesn’t change our generic debugging steps, but it means that we may want to take some extra notes in a text editor for crafting a prompt if necessary. 

Author: ```suggestion
Adding in ChatGPT doesn’t change our generic debugging steps, but it means that we may want to take some extra notes in a text editor for crafting a prompt if necessary. We should try our best to provide as many details as possible so that ChatGPT does not need to fill in gaps or make assumptions which might result in an incorrect response. 
```

As we investigate what is happening, we should take notes of the exact issues our code is experiencing such as: 
- compiler errors that the IDE generates 
- runtime exceptions that occur when the code is executing
- any other unexpected behavior in the code that could stem from a logical error.

When we ask ChatGPT for assistance, we need to be descriptive to get useful help. Our prompts for analyzing or explaining code should include:
- a detailed description of what the code is meant to do
- a detailed description of the input, expected result, and actual result
- any other information needed to reliably reproduce the issue being debugged

> The type of info we’re gathering likely sounds familiar - when we ask ChatGPT for help we need to gather the same kinds of detailed information and explanation that we would use if we were reaching out to other people for help who may be unfamiliar with the project. 

Reviewer: Think you wrote out information everywhere else, so updating 'info' to 'information' would be more consistent.

Also, how would you feel about adding a sentence like, "We should try our best to provide as many details as possible so that ChatGPT does not need to fill in gaps or make assumptions' which might result in an incorrect response"?

Author: I really like that added line, and thanks for catching the info/information inconsistency!

> The type of info we’re gathering likely sounds familiar - when we ask ChatGPT for help we need to gather the same kinds of detailed information and explanation that we would use if we were reaching out to other people for help who may be unfamiliar with the project. 

Author: ```suggestion
The type of information we’re gathering likely sounds familiar - when we ask ChatGPT for help we need to collect the same kinds of detailed information and explanation that we would use if we were reaching out to other people for help who may be unfamiliar with the project. 
```

> 18.         if self.condition > 3:
> 19.             return "It's great!"
> 20.         else:
> 21.             return "It stinks!"
> ```
> 
> **Code Block 2: Clothing Class**
> ```python
> 23. from swap_meet.item import Item
> 24. 
> 25. class Clothing(Item):
> 26.     def __init__(self, condition=None, id=None, fabric="Unknown"):
> 27.         super().__init__(condition=condition, id=id)
> 28.         self.fabric = fabric
> 29. 
> 30.     def __str__(self):
> 31.         return f"{super()._str_()} It is made from {self.condition} fabric."
> ```
>
> I've included line numbers for reference. If you have any questions or need further clarification, feel free to ask!

</details>  

## Adding ChatGPT to the Debugging Toolbox

The key to effectively using ChatGPT for debugging is that we’re adding a new tool to our toolbox, not replacing the ones we’ve already built. The general steps of our debugging journey should look pretty much the same: we want to determine what is happening, why it’s happening, and how we fix it. 

Let’s take a closer look at the debugging journey to see where we can make the most use of tools like ChatGPT.

### What’s Happening

#### Investigation

> When we’re debugging the problem may not always be obvious, so we may have to do some searches about errors, set breakpoints, or add print statements to better understand what kinds of inputs are giving us an issue and how. If our initial searches don’t come up with clear answers, we can ask ChatGPT questions about the error messages or the behavior we are seeing to get a better idea of what is causing the problem and where it may stem from.

Reviewer: ```suggestion
When we’re debugging the problem may not always be obvious, so we may have to do some searches about errors, set breakpoints, or add print statements to better understand what kinds of inputs are giving us an issue and why. If our initial searches don’t come up with clear answers, we can ask ChatGPT questions about the error messages or the behavior we are seeing to get a better idea of what is causing the problem.
```

Reviewer: Sentence felt a little long so I feel like the last bit can be cut but not a big deal if you'd like to leave it in.

### Why is it Happening?

#### Understand the Code

We need a solid understanding of how a piece of code works to debug effectively. As we identified what was happening we may have gotten an understanding of the code, but some projects are complex or need a bit more time to study. If we want a clearer picture of the overall code, we can set a timebox for reading through the code line by line, rubber ducking, charting execution, or any other of our prior tools you find helpful. If we hit our timebox, ChatGPT may be able to help us understand the code by answering questions about its purpose, structure, and syntax.

#### Isolate & Reproduce the Problem

Next we want to dig into why the bug is happening. We want to isolate the problem by finding the line or lines of code causing our issue and reliably reproduce the bug so we know how to test for it and confirm the bug was fixed. Our prior tools are a great place to start, stepping through our code or charting the execution to see when unexpected values begin to appear. 

> A helpful strategy can be to set a timebox for stepping through the code with a debugger and if that ends, then reach out to ChatGPT for help determining which lines of code are causing the issue. ChatGPT can also help us use other debugging tools more effectively if we ask for guidance. For example, if we pinpoint where the bug is happening after having trouble doing so with a debugger, we can ask ChatGPT “How could we catch the issue with a debugger?”.

Reviewer: ```suggestion
A helpful strategy can be to set a timebox for stepping through the code with a debugger and when that ends, then reach out to ChatGPT for help determining which lines of code are causing the issue. ChatGPT can also help us use other debugging tools more effectively if we ask for guidance. For example, if we pinpoint where the bug is happening after having trouble doing so with a debugger, we can ask ChatGPT “How could we catch the issue with a debugger?”.
```

> Similarly, if we’re having trouble reproducing the bug consistently, we can set a timebox for exploration and research. If the timebox ends, we can see if ChatGPT can help guide us on how to reproduce the issue and suggest test cases to let us reliably test the issue.

Reviewer: ```suggestion
Similarly, if we’re having trouble reproducing the bug consistently, we can set a timebox for exploration and research. If the timebox ends and we still don't have our answer, we can see if ChatGPT can help guide us on how to reproduce the issue and suggest test cases to let us reliably test the issue.
```

>
> Here's the updated code:
> 
> ```python
> import random
> 
> def draw_letters():
>     distribution = {'A': 9, 'B': 2, 'O': 8, 'C': 2, 'P': 2, 
>                     'D': 4, 'Q': 1, 'E': 12, 'R': 6, 'F': 2, 
>                     'S': 4, 'G': 3, 'T': 6, 'H': 2, 'U': 4, 
>                     'I': 9, 'V': 2, 'J': 1, 'W': 2, 'K': 1, 
>                     'X': 1, 'L': 4, 'Y': 2, 'M': 2, 'Z': 1}
>     
>     pool = []
>     for letter in distribution:
>         pool.extend([letter] * distribution[letter])
> 
>     hand = random.sample(pool, 10)
> 
>     return hand
> ```
> 
> This code uses `random.sample(pool, 10)` to directly obtain a list of 10 unique elements from the `pool` list, eliminating the need for the loop and manual removal of elements.

We can see that ChatGPT is suggesting that Althea uses a separate function `sample` from the `random` module. Before copy/pasting the suggestion into their project, they go back to the random module documentation or other trusted resources to ensure: 
- Does the `random.sample` function exist? 

> - If `random.sample` does exist, does it actually do what ChatGPT thinks it does? 

Reviewer: ```suggestion
- If `random.sample` does exist, does it actually do what ChatGPT says it does? 
```

> - If it does what ChatGPT thinks, then does using the function meet the original intention of the code we’re trying to fix?

Reviewer: ```suggestion
- If it does what ChatGPT thinks, then does using the function meet the original code requirements?
```

In this case, the answer to all of these questions is Yes! Althea can once again update their code and run the test suite to ensure everything is still in working order after their changes 🎉.

## Summary

> ChatGPT can be a really helpful tool to give us more insight on what code is doing, but we need to be careful about over-reliance. There are many situations in life such as interviews and code review meetings where we may not have access to generative AI tools, meaning that tools like ChatGPT are best used to support and augment the debugging skills we’ve built rather than to replace them. When we start out a debugging journey, we still want to begin with searching errors on the web, using print statements, and setting breakpoints to step through our code. 

Reviewer: Could we add 'asking others for help' too?

"When we start out a debugging journey, we still want to begin with searching errors on the web, using print statements, setting breakpoints to step through our code, and asking others for help before turning to ChatGPT."

> ChatGPT can be a really helpful tool to give us more insight on what code is doing, but we need to be careful about over-reliance. There are many situations in life such as interviews and code review meetings where we may not have access to generative AI tools, meaning that tools like ChatGPT are best used to support and augment the debugging skills we’ve built rather than to replace them. When we start out a debugging journey, we still want to begin with searching errors on the web, using print statements, and setting breakpoints to step through our code. 

Author: ```suggestion
ChatGPT can be a really helpful tool to give us more insight on what code is doing, but we need to be careful about over-reliance. There are many situations in life such as interviews and code review meetings where we may not have access to generative AI tools, meaning that tools like ChatGPT are best used to support and augment the debugging skills we’ve built rather than to replace them. When we start out a debugging journey, we still want to begin with searching errors on the web, using print statements, setting breakpoints to step through our code, and asking others for help before turning to ChatGPT. 
```

> If we have trouble finding resources about the error or understanding what the code is doing using our prior skills, we can begin to augment our investigation by folding in ChatGPT to see if it can explain an error and common causes, or talk through what particular lines of code are doing in reality vs expectation. 

Reviewer: https://tenor.com/view/schitts-creek-david-moira-cheese-fold-gif-14734408 😄

The more we work with code, the more familiar we will be with common edge cases and coming up with the edge cases unique to our project, but we can also ask ChatGPT to help us brainstorm test cases to help strengthen our test suites. 

## Check for Understanding

<!-- Question 1 -->
<!-- prettier-ignore-start -->
### !challenge
* type: ordering
* id: f22a6178-953d-11ee-b9d1-0242ac120002
* title: Ordering Althea's debugging steps
##### !question

Althea is debugging another project that is producing a `TypeError` and they are trying to pinpoint the line causing them trouble. 
Order the debugging steps that they took from first to last action:

##### !end-question
##### !answer

1. Do a web search for the error
1. Go line by line and rubber duck
1. Use a debugger to step through the code
1. Prompt ChatGPT for assistance pinpointing the bug
1. Step through the code to confirm if the line ChatGPT suggested is where unexpected behavior starts to appear

##### !end-answer
##### !explanation

> Althea wants to first do some quick searches for the error to see if there are resources on what the error means and why it might be happening. They want to understand the code better, so they do a close reading of the code and do some rubber ducking. Once they feel more clear on the overall goal of the code, they set some breakpoints and step through the code line by line to try to suss out where the strange behavior starts. Althea has an idea of what is going wrong and has narrowed down the issue to a few lines of code but they want more clarity, so they write up what they learned from their observations and prompt ChatGPT for guidance on where the issue might be happening. After carefully reading ChatGPT’s response to ensure it makes sense, they step through their code again, paying close attention to the line that ChatGPT called out as a potential issue and ensuring they see the the bug happening at that line.

Reviewer: ```suggestion
Althea wants to first do some quick searches for the error to see if there are resources on what the error means and why it might be happening. They want to understand the code better, so they do a close reading of the code and do some rubber ducking. Once they feel more clear on the overall goal of the code, they set some breakpoints and step through the code line by line to try to suss out where the strange behavior starts. Althea has an idea of what is going wrong and has narrowed down the issue to a few lines of code but they want more clarity, so they write up what they learned from their observations and prompt ChatGPT for guidance on where the issue might be happening. After carefully reading ChatGPT’s response to ensure it makes sense, they step through their code again, paying close attention to the line that ChatGPT called out as a potential issue and double check that they see the the bug happening at that line.
```

##### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 2 -->
<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: e3d729c2-9541-11ee-b9d1-0242ac120002
* title: Sharing code with ChatGPT

##### !question

Take a look at [the snowbug repo](https://github.com/AdaGold/snowbug), particularly the source files [main.py](https://github.com/AdaGold/snowbug/blob/main/main.py) and [snowman/snowman.py](https://github.com/AdaGold/snowbug/blob/main/snowman/snowman.py). 

For this problem, let’s say that we’ve fixed all the bugs in `snowman.py` except for one! The losing message always prints at the correct time, but there is something wrong with the string that’s printed. We stepped through the code and are sure there is an issue with the losing message on line 43 of the `snowman` function in the file `snowman/snowman.py`. 

```
print(f"Sorry, you lose! The word was {snowman}")
```

> If we wanted to craft a prompt to help debug this specific issue, what is the least amount of code we could share with a LLM tool that would give it enough context to help us debug? Choose one option below:

Reviewer: Can we bold "least" for emphasis?

> If we wanted to craft a prompt to help debug this specific issue, what is the least amount of code we could share with a LLM tool that would give it enough context to help us debug? Choose one option below:

Author: ```suggestion
If we wanted to craft a prompt to help debug this specific issue, what is the *least* amount of code we could share with a LLM tool that would give it enough context to help us debug? Choose one option below:
```
